File: src_zheqiang/parameters.f90
! Read model parameters
module m_parameters
implicit none
contains

subroutine read_parameters
use m_globals
use m_fieldio
integer :: ios, i
character(12) :: key
character(1) :: op
character(256) :: line

! I/O pointers
allocate( io0 )
io => io0
io%next => io0
io%field = 'head'

open( 1, file='parameters.py', status='old' )

doline: do

! Read line
read( 1, '(a)', iostat=ios ) line
if ( ios /= 0 ) exit doline

! Strip comments and punctuation
str = line
i = scan( str, '#' )
if ( i > 0 ) str(i:) = ' '
do
    i = scan( str, "()[]{}'," )
    if ( i == 0 ) exit
    str(i:i) = ' '
end do

! Read key val pair
if ( str == '' ) cycle doline
read( str, *, iostat=ios ) key

! Select input key
select case( key )
case( 'fieldio', '' )
case( 'nn' );           read( str, *, iostat=ios ) key, op, nn
case( 'nt' );           read( str, *, iostat=ios ) key, op, nt
case( 'dx' );           read( str, *, iostat=ios ) key, op, dx
case( 'dt' );           read( str, *, iostat=ios ) key, op, dt
case( 'tm0' );          read( str, *, iostat=ios ) key, op, tm0
case( 'affine' );       read( str, *, iostat=ios ) key, op, affine
case( 'n1expand' );     read( str, *, iostat=ios ) key, op, n1expand
case( 'n2expand' );     read( str, *, iostat=ios ) key, op, n2expand
case( 'rexpand' );      read( str, *, iostat=ios ) key, op, rexpand
case( 'gridnoise' );    read( str, *, iostat=ios ) key, op, gridnoise
case( 'oplevel' );      read( str, *, iostat=ios ) key, op, oplevel
case( 'rho1' );         read( str, *, iostat=ios ) key, op, rho1
case( 'rho2' );         read( str, *, iostat=ios ) key, op, rho2
case( 'vp1' );          read( str, *, iostat=ios ) key, op, vp1
case( 'vp2' );          read( str, *, iostat=ios ) key, op, vp2
case( 'vs1' );          read( str, *, iostat=ios ) key, op, vs1
case( 'vs2' );          read( str, *, iostat=ios ) key, op, vs2
case( 'gam1' );         read( str, *, iostat=ios ) key, op, gam1
case( 'gam2' );         read( str, *, iostat=ios ) key, op, gam2
case( 'vdamp' );        read( str, *, iostat=ios ) key, op, vdamp
case( 'hourglass' );    read( str, *, iostat=ios ) key, op, hourglass
case( 'bc1' );          read( str, *, iostat=ios ) key, op, bc1
case( 'bc2' );          read( str, *, iostat=ios ) key, op, bc2
case( 'npml' );         read( str, *, iostat=ios ) key, op, npml
case( 'i1pml' );        read( str, *, iostat=ios ) key, op, i1pml
case( 'i2pml' );        read( str, *, iostat=ios ) key, op, i2pml
case( 'ppml' );         read( str, *, iostat=ios ) key, op, ppml
case( 'vpml' );         read( str, *, iostat=ios ) key, op, vpml
case( 'ihypo' );        read( str, *, iostat=ios ) key, op, ihypo
case( 'source' );       read( str, *, iostat=ios ) key, op, source
case( 'timefunction' ); read( str, *, iostat=ios ) key, op, timefunction
case( 'period' );       read( str, *, iostat=ios ) key, op, period
case( 'source1' );      read( str, *, iostat=ios ) key, op, source1
case( 'source2' );      read( str, *, iostat=ios ) key, op, source2
case( 'nsource' );      read( str, *, iostat=ios ) key, op, nsource
case( 'ivols' );        read( str, *, iostat=ios ) key, op, ivols
case( 'eplasticity' );  read( str, *, iostat=ios ) key, op, eplasticity
case( 'tv' );           read( str, *, iostat=ios ) key, op, tv
case( 'friction' );     read( str, *, iostat=ios ) key, op, friction
case( 'pcdep' );        read( str, *, iostat=ios ) key, op, pcdep
case( 'faultnormal' );  read( str, *, iostat=ios ) key, op, faultnormal
case( 'slipvector' );   read( str, *, iostat=ios ) key, op, slipvector
case( 'faultopening' ); read( str, *, iostat=ios ) key, op, faultopening
case( 'vrup' );         read( str, *, iostat=ios ) key, op, vrup
case( 'rcrit' );        read( str, *, iostat=ios ) key, op, rcrit
case( 'trelax' );       read( str, *, iostat=ios ) key, op, trelax
case( 'rnucl' );        read( str, *, iostat=ios ) key, op, rnucl
case( 'tmnucl' );       read( str, *, iostat=ios ) key, op, tmnucl
case( 'delts' );        read( str, *, iostat=ios ) key, op, delts
case( 'svtol' );        read( str, *, iostat=ios ) key, op, svtol
case( 'np3' );          read( str, *, iostat=ios ) key, op, np3
case( 'itstats' );      read( str, *, iostat=ios ) key, op, itstats
case( 'itio' );         read( str, *, iostat=ios ) key, op, itio
case( 'itcheck' );      read( str, *, iostat=ios ) key, op, itcheck
case( 'itstop' );       read( str, *, iostat=ios ) key, op, itstop
case( 'debug' );        read( str, *, iostat=ios ) key, op, debug
case( 'mpin' );         read( str, *, iostat=ios ) key, op, mpin
case( 'mpout' );        read( str, *, iostat=ios ) key, op, mpout
case default
    select case( key(1:1) )
    case( '=', '+' )
        call pappend
        io%ib = -1
        !XXXread( str, *, iostat=ios ) io%mode, io%nc
        read( str, *, iostat=ios ) io%mode, io%nc, io%tfunc, &
            io%period, io%x1, io%x2, io%nb, io%ii, io%filename, &
            io%val, io%field
    case default; ios = 1
    end select
end select

! Error check
if ( ios /= 0 ) then
    if ( master ) write( 0, * ) 'bad input: ', trim( line )
    stop
end if

end do doline

close( 1 )

end subroutine

end module

File: parameters.py
# -*- coding: utf-8 -*-
# @Author: yow004
# @Date:   2018-08-18 01:41:21
# @Last Modified by:   yow004
# @Last Modified time: 2018-08-23 15:29:21
#!/usr/bin/env python
"""
Default simulation parameters

Spatial difference operator level:

  0: Auto pick 2 or 6
  1: Mesh with constant spacing dx
  2: Rectangular mesh
  3: Parallelepiped mesh
  4: One-point quadrature
  5: Exactly integrated elements
  6: Saved operators, nearly as fast as 2, but doubles the memory usage
"""

# I/O and code execution parameters
np3 = 1, 1, 1			# number of processors in j k l
mpin = 1			# input:  0=separate files, 1=MPI-IO, -1=non-collective MPI-IO
mpout = 1			# output: 0=separate files, 1=MPI-IO, -1=non-collective MPI-IO
itstats = 10			# interval for calculating statistics
itio = 50			# interval for writing i/o buffers
itcheck = 0			# interval for check-pointing (0=off)
itstop = 0			# for testing check-pointing, simulates code crash
debug = 0			# >0 verbose, >1 sync, >2 mpi vars, >3 I/O

# Wave model parameters
nn = 41, 41, 42			# number of nodes in j, k, l (double nodes counted)
nt = 41				# number of time steps
dx = 100.0, 100.0, 100.0	# spatial step length
dt = 0.0075			# time step length
tm0 = 0.0			# initial time
affine = (1.0, 0.0, 0.0), (0.0, 1.0, 0.0), (0.0, 0.0, 1.0) # grid transformation
gridnoise = 0.0			# random noise added to mesh, assumes planar fault
oplevel = 0			# spatial difference operator level
vdamp = -1.0			# Vs dependent damping
hourglass = 1.0, 1.0		# hourglass stiffness (1) and viscosity (2)
fieldio = [			# field I/O
    ('=', 'rho', [], 2670.0),	# density
    ('=', 'vp',  [], 6000.0),	# P-wave speed
    ('=', 'vs',  [], 3464.0),	# S-wave speed
    ('=', 'gam', [],    0.0),	# viscosity
    ('=', 'mco', [],    0.0),   # material cohesion
]
rho1 = -1.0			# min density
rho2 = -1.0			# max density
vp1 = -1.0			# min P-wave speed
vp2 = -1.0			# max P-wave speed
vs1 = -1.0			# min S-wave speed
vs2 = -1.0			# max S-wave speed
gam1 = -1.0			# min viscosity
gam2 = 0.8			# max viscosity
npml = 10			# number of PML damping nodes
ppml = 2			# PML exponend, 1-4. Generally 2 is best.
vpml = -1.0			# PML damping velocity, <0 default to min, max V_s harmonic mean
bc1 = 0, 0, 0			# boundary condition - near side
bc2 = 0, 0, 0			# boundary condition - far side
ihypo = 0, 0, 0			# hypocenter indices (with fractional values), 0 = center
rexpand = 1.06			# grid expansion ratio
n1expand = 0, 0, 0		# number of grid expansion nodes - near side
n2expand = 0, 0, 0		# number of grid expansion nodes - far side

# Dynamic rupture parameters
faultnormal = 0			# normal direction to fault plane (0=no fault)
faultopening = 0		# 0=not allowed, 1=allowed
slipvector = 1.0, 0.0, 0.0	# shear traction direction for ts1
svtol = 0.001			# slip velocity considered rupturing

# nucleation by gradually overstressing
vrup = -1.0                     # nucleation rupture velocity, negative = no nucleation
rcrit = 0.0                     # nucleation critical radius
trelax = -1.0                   # nucleation relaxation time (0.07sec)
rrelax = 100					# prescribed cohesive zone size (100 m by default)
tslope = -1.0                   # nucleation relaxation slope (choose trelax or tslope) (5MPa/km)
rnucl = 0.0                     # nucleation radius/half-width for Ts perturbation
tmnucl = 1.0                    # duration time for Ts pertubation. if negative, perturb at t=0
delts = 0.0                     # peak amplitude increase for Ts perturbation
psidelts = -1.0                 # peak amplitude reduction ratio for psi perturbation
nstage = 1						# used in forced rupture. Numbers of stages of different rupture velocity
vrupstage = 3300				# rupture velocity in each segment vrup = 3000, 3500, 5000
sizestage = 2000				# each segment size

# friction type
friction = 'slipweakening'      # or 'rateandstate' 'frictionless (need input sa)' 'forced (need rrelax)'
intype = 'sa'			  # or 'sv' or 'sa' this defines kinematic input type slip rate or slip acc (time history)
					  # fsu is input static final slip to calculate static stress change (tiny density)
					  # period is used for artificial period static slip input

# Prakash-Clifton gradual response to change of normal traction
pcdep = 'no'                    # or 'yes'  

# elastic or plastic
eplasticity = 'elastic'         # or 'elastoplastic' 
tv = -1.0                       # viscoplastic relaxation time Tv = dx/Vs = 50./3464.

# frequency-dependent attenuation
attenuation = 'elastic'         # or 'anelastic'
fac = 1.0			# factor to shift bandwith for constant Q only (f0 in paper)
q0 = 150.0			# Q0 value in Q0*f^n relation (not directly used here)
ex = 0.0 			# Exp of powerlaw (use 0.1:0.1:0.9 for Q(f),0.0 for constant
fp = 1.0			# reference frequency (velocity of input media)

# Skempton's coefficient in pore elastic medium
#skepb = 0.0                     # skepb in (0,1) 0.0 means no pore pressure change, 
                                # 1.0 means effective normal traction keeps constant

# initial volume stress input 
ivols = 'no'

# Finite source parameters
source = 'none'		# 'moment', 'potency', 'force','fault', or 'none'
nsource = 0			# number of sub-faults

# Point source parameters
source1 = 0.0, 0.0, 0.0		# normal components
source2 = 0.0, 0.0, 0.0		# shear components
strike = 90.
dip = 90.
rake = 0.
m0 = 1.e16                # N.m (Mw4.6)
timefunction = 'none'		# time function, see util.f90 for details.
period = 10 * dt		# dominant period

# Placeholders
i1pml = None
i2pml = None

